Simple FTP is done

viktoriia-fomina opened the pull request.

viktoriia-fomina committed:
Created a project

viktoriia-fomina committed:
Renamed folder

viktoriia-fomina committed:
i've started implementation of the server

viktoriia-fomina committed:
i've made structure of the projects better

viktoriia-fomina committed:
Server's ready. RequestHandler is not implemented yet

viktoriia-fomina committed:
Server and RequestHandler are done

viktoriia-fomina committed:
Client is implemented and test project is created. Uselles usings are removed

viktoriia-fomina committed:
Added tests, ClientNotRunningException class

yurii-litvinov commented on SimpleFTP/SimpleFTPClient/SimpleFTPClient/Client.cs:
```diff
@@ -0,0 +1,96 @@
+    /// Client that can make two requests: List (listing files in the server's directory) and Get (downloading a
+    /// file from the server's directory).
+    /// </summary>
+    public class Client
```
TcpClient реализует IDisposable, этот класс хранит его в качестве поля, значит и сам должен реализовывать IDisposable, и вызывать Dispose() у TcpClient

yurii-litvinov commented on SimpleFTP/SimpleFTPClient/SimpleFTPClient/Client.cs:
```diff
@@ -0,0 +1,96 @@
+        /// </summary>
+        public void Run()
+        {
+            client = new TcpClient(server, port);
```
Это открывает соединение, потенциально длительная операция, так что стоило бы использовать [TcpClient.ConnectAsync](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.tcpclient.connectasync?view=net-5.0)

yurii-litvinov commented on SimpleFTP/SimpleFTPClient/SimpleFTPClient/Client.cs:
```diff
@@ -0,0 +1,96 @@
+                throw new FileNotFoundException("File was not found on the server!");
+            }
+
+            using (var fileStream = new FileStream(downloadTo + fileName, FileMode.CreateNew))
```
Можно было using var, без скобок

yurii-litvinov commented on SimpleFTP/SimpleFTPClient/SimpleFTPClient/Client.cs:
```diff
@@ -0,0 +1,96 @@
+
+            await writer.WriteLineAsync("1" + path);
+
+            return await reader.ReadLineAsync();
```
Возвращать строку не очень интересно, хочется более высокоуровневую модель данных, с которой потом было бы приятно работать.

yurii-litvinov commented on SimpleFTP/SimpleFTPClient/SimpleFTPClient/Client.cs:
```diff
@@ -0,0 +1,96 @@
+        {
+            if (client == null || !client.Connected)
+            {
+                throw new ClientNotRunningException("Client is not running!");
```
Message тут несколько избыточен, по имени исключения можно понять, что произошло.

yurii-litvinov commented on SimpleFTP/SimpleFTPServer/SimpleFTPServer/Server.cs:
```diff
@@ -0,0 +1,62 @@
+        {
+            Task.Run(async () =>
+            {
+                while (!cancellationToken.IsCancellationRequested)
```
Где-то тут надо брать using на client, иначе его никто никогда не закроет

yurii-litvinov commented on SimpleFTP/SimpleFTPServer/SimpleFTPServer/Server.cs:
```diff
@@ -0,0 +1,62 @@
+        /// <summary>
+        /// Stops the server.
+        /// </summary>
+        public void Stop() => listener.Stop();
```
Тоже, стоит Server сделать IDisposable

yurii-litvinov commented on SimpleFTP/SimpleFTPServer/SimpleFTPServer/RequestHandler.cs:
```diff
@@ -0,0 +1,131 @@
+                return;
+            }
+
+            var root = Directory.GetParent(Directory.GetParent(Directory.GetParent(Directory.GetCurrentDirectory()).Parent.FullName).Parent.FullName).FullName;
```
Это, наверное, как-то проще можно, типа "./../.."

yurii-litvinov commented on SimpleFTP/SimpleFTPServer/SimpleFTPTests/SimpleFTPTests.cs:
```diff
@@ -0,0 +1,103 @@
+
+            server.Stop();
+
+            client.Close();Assert.AreEqual("1 .\\Test\\Folder\\text.txt false ", listRequestFolderResult);
```
Перевода строки не хватает :)

yurii-litvinov commented on SimpleFTP/SimpleFTPServer/SimpleFTPTests/SimpleFTPTests.cs:
```diff
@@ -0,0 +1,103 @@
+                File.Delete(pathToFile);
+
+                client.Close();
+            });
```
А Вы уверены, что до окончания теста этот Task исполнится? А зачем он, кстати?

viktoriia-fomina committed:
Fixed part of remarks:

- I've made a client IDisposable
- In requestHandler since now I better look for path

viktoriia-fomina committed:
Client's response since now is not a string but a List of pairs (string, bool) - more high-level model)

viktoriia-fomina committed:
added fixes:

- Made client and server IDisposable

yurii-litvinov commented on SimpleFTP/SimpleFTPClient/SimpleFTPClient/Client.cs:
```diff
@@ -0,0 +1,131 @@
+        /// Does listing.
+        /// </summary>
+        /// <exception cref="ClientNotRunningException">Thrown when the client is not running.</exception>
+        public async Task<List<(string, bool)>> List(string path)
```
Так лучше, но если бы он список структур с именованными полями возвращал, было бы ещё лучше :)

yurii-litvinov commented on SimpleFTP/SimpleFTPClient/SimpleFTPClient/Client.cs:
```diff
@@ -0,0 +1,131 @@
+        /// <exception cref="DirectoryNotFoundException">Thrown when request directory was not found.</exception>
+        private List<(string, bool)> ParseListResponse(string response)
+        {
+            var splitedResponse = response.Split(' ');
```
split, не splited

yurii-litvinov commented on SimpleFTP/SimpleFTPClient/SimpleFTPClient/ClientNotRunningException.cs:
```diff
@@ -0,0 +1,16 @@
+namespace SimpleFTPClient
+{
+    /// <summary>
+    /// The exception that is thrown when the client is not running.
```
Непрерывно бросается, пока клиент не запустят? :)

yurii-litvinov commented on SimpleFTP/SimpleFTPClient/SimpleFTPClient/ClientNotRunningException.cs:
```diff
@@ -0,0 +1,16 @@
+        /// <summary>
+        /// Initializes a new instance of the ClientNotRunningException class.
+        /// </summary>
+        public ClientNotRunningException() : base() { }
```
А тогда этот конструктор можно не писать, он в классе один и совпадает с конструктором по умолчанию.

yurii-litvinov commented on SimpleFTP/SimpleFTPServer/SimpleFTPServer/Program.cs:
```diff
@@ -0,0 +1,11 @@
+    {
+        public static void Main(string[] args)
+        {
+
```
Ну, кстати, неплохо бы иметь консольный раннер, запускающий сервер

yurii-litvinov reviewed (approved):
Окей, в целом всё так, зачтена.

viktoriia-fomina committed:
Renamed: splitedResponse -> splitResponse. ClientNonRunningException comment is more logical now
